### PR TITLE
[FLINK-12959][table-planner-blink] Use BoundedInput and InputSelectable in blink and implement hash join

### PR DESCRIPTION
**Pull request by JingsongLi**
## What is the purpose of the change

1.Use BoundedInput and InputSelectable in blink
2.Implement BatchExecHashJoin translateToPlanInternal

## Verifying this change

ut

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)

### CODE REVIEW

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/plan/nodes/physical/batch/BatchExecExchange.scala [hunk 1]
```diff
@@ -163,37 +163,45 @@ class BatchExecExchange(
...
 
-    // TODO supports DataExchangeMode.BATCH in runtime
-    if (requiredExchangeMode.contains(DataExchangeMode.BATCH)) {
-      throw new TableException("DataExchangeMode.BATCH is not supported now")
+    val shuffleMode = requiredExchangeMode match {
+      case None => ShuffleMode.PIPELINED
+      case Some(mode) =>
+        mode match {
+          case DataExchangeMode.BATCH => ShuffleMode.BATCH
+          case DataExchangeMode.PIPELINED => ShuffleMode.PIPELINED
```
Comment (liyafan82) [inline flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/plan/nodes/physical/batch/BatchExecExchange.scala]: how about PIPELINE_WITH_BATCH_FALLBACK?

File: flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/api/TableConfigOptions.java [hunk 2]
```diff
@@ -151,6 +151,16 @@
 					.defaultValue(32)
 					.withDescription("Sets the table reserved memory size of hashAgg operator. It defines the lower limit.");
 
+	public static final ConfigOption<Integer> SQL_RESOURCE_HASH_JOIN_TABLE_MEM =
+			key("sql.resource.hash-join.table.memory.mb")
+					.defaultValue(32)
+					.withDescription("Sets the HashTable reserved memory for hashJoin operator. It defines the lower limit for.");
```
Comment (liyafan82) [inline flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/api/TableConfigOptions.java]: remove "for"

File: flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/join/SortMergeJoinOperator.java [hunk 3]
```diff
@@ -209,17 +210,21 @@ public void processElement2(StreamRecord<BaseRow> element) throws Exception {
...
+				if (isAllFinished()) {
+					doSortMergeJoin();
+				}
+				break;
+			case 2:
+				isFinished[1] = true;
+				if (isAllFinished()) {
+					doSortMergeJoin();
+				}
+				break;
```
Comment (liyafan82) [inline flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/join/SortMergeJoinOperator.java]: The code seems duplicated. 
Maybe you can replace with 

isFinished[inputId] = true;
if (isAllFinished()) {
    doSortMergeJoin();
}

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/plan/nodes/physical/batch/BatchExecExchange.scala (hunk 1 above)
Comment (JingsongLi) [reply flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/plan/nodes/physical/batch/BatchExecExchange.scala]: Now, table level never produce a `PIPELINE_WITH_BATCH_FALLBACK`.
There is a exception when `PIPELINE_WITH_BATCH_FALLBACK`.

File: flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/join/SortMergeJoinOperator.java (hunk 3 above)
Comment (JingsongLi) [reply flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/join/SortMergeJoinOperator.java]: `isFinished[inputId - 1] = true`

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/codegen/CodeGenUtils.scala [hunk 4]
```diff
@@ -20,6 +20,7 @@ package org.apache.flink.table.codegen
 
 import org.apache.flink.api.common.ExecutionConfig
 import org.apache.flink.core.memory.MemorySegment
+import org.apache.flink.table
```
Comment (KurtYoung) [inline flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/codegen/CodeGenUtils.scala]: useless import?

File: flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/api/TableConfigOptions.java [hunk 5]
```diff
@@ -151,6 +151,11 @@
 					.defaultValue(32)
 					.withDescription("Sets the table reserved memory size of hashAgg operator. It defines the lower limit.");
 
+	public static final ConfigOption<Integer> SQL_RESOURCE_HASH_JOIN_TABLE_MEM =
+			key("sql.resource.hash-join.table.memory.mb")
+					.defaultValue(32)
```
Comment (KurtYoung) [inline flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/api/TableConfigOptions.java]: seems too small

Comment (JingsongLi) [reply flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/api/TableConfigOptions.java]: Now hash agg and sort default memory is 32M too.

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/codegen/CodeGenUtils.scala (hunk 4 above)
Comment (JingsongLi) [reply flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/codegen/CodeGenUtils.scala]: OK

**Pull request closed without merging**